### PR TITLE
python(feature): file attachments upload + download service

### PR DESCRIPTION
**Pull request by solidiquis**
## Changes

Added `sift_py.file_attachment` module to leverage the file attachments API. The `FileAttachmentService` from said module allows for ergonomic uploading, downloading, retrieving attachments for various entities, as well as bulk deleting.

Unit testing + documentation updates included. Examples will be added in a separate PR, though the documentation already includes plenty of examples.

 
Screenshot 2024-08-16 at 16 38 39

### CODE REVIEW

**Comment by zachAtSiftStack** [comment]
nit: docs typo, drop the word "support"
Screenshot 2024-08-19 at 11 15 51 AM


> it also support provides utilities....

File: python/lib/sift_py/file_attachment/__init__.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+from sift_py.file_attachment.service import FileAttachmentService
+from sift_py.file_attachment.entity import Entity, EntityType
+from sift_py.file_attachment.metadata import ImageMetadata
+from sift_py.rest import SiftRestConfig
+
+from sift.remote_files.v1.remote_files_pb2 import GetRemoteFileRequest
+from sift.remote_files.v1.remote_files_pb2_grpc import RemoteFileServiceStub
+
+rest_config: SiftRestConfig = {
+    # Be sure to exclude the "https://" or "http://" scheme out of the uri
```
Comment (zachAtSiftStack) [inline python/lib/sift_py/file_attachment/__init__.py:25]: Mainly curious. Any reason we don't just give the prod / gov examples for the rest endpoints? If it is a security or privacy thing? I am also fine with it as is

Comment (solidiquis) [reply python/lib/sift_py/file_attachment/__init__.py:25]: I fixed the typo that's still in the screenshot. Screenshot is out of date. And yeah it's for security purposes. Something we only want to expose via documentation to customers. It's a minor inconvenience though for sure.

**Comment by zachAtSiftStack** [comment]
The error log for a bad api key or run id (not sure which) wasn't really helpful

```
(filevenv) (venv) ➜  ingestion_with_python_config git:(python/feature-file-attachment) ✗ python3 test_file_attach.py
Traceback (most recent call last):
  File "/Users/zachplunkett/Code/sift/python/examples/ingestion_with_python_config/test_file_attach.py", line 45, in <module>
    remote_file = file_attachment_service.upload_attachment(
  File "/Users/zachplunkett/venvs/filevenv/lib/python3.9/site-packages/sift_py/file_attachment/service.py", line 80, in upload_attachment
    remote_file_id = self._upload_service.upload_attachment(
  File "/Users/zachplunkett/venvs/filevenv/lib/python3.9/site-packages/sift_py/file_attachment/_internal/upload.py", line 90, in upload_attachment
    raise Exception(response.text)
Exception
```

But after that it worked as expected :)

Screenshot 2024-08-19 at 6 03 53 PM

File: python/lib/sift_py/file_attachment/_internal/upload.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        base_uri = self.__class__._compute_uri(restconf)
+        self._upload_uri = urljoin(base_uri, self.UPLOAD_PATH)
+        self._upload_bulk_uri = urljoin(base_uri, self.UPLOAD_BULK_PATH)
+        self._apikey = restconf["apikey"]
+
+    def upload_attachment(
+        self,
+        path: Union[str, Path],
+        entity: Entity,
+        metadata: Optional[Metadata],
```
Comment (zachAtSiftStack) [inline python/lib/sift_py/file_attachment/_internal/upload.py]: This is still required but marked optional

**Review by zachAtSiftStack — APPROVED** [review]
LGTM, Approving

File: python/lib/sift_py/file_attachment/__init__.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    run = entity=Entity(
+        entity_id=run_id, # some arbitrary run ID that refers to an existing run
+        entity_type=EntityType.RUN,
+    )
+
+    # uploading the file attachment and attaching it to a run of `run_id`
+    remote_file = file_attachment_service.upload_attachment(
+        path="path/to/foo.mp4",
+        entity=run,
+        # Metatadata.. optional but recommended for optimal viewing in the application
```
[the remark this answers is not in the document]
Comment (zachAtSiftStack) [reply python/lib/sift_py/file_attachment/__init__.py:70]: Addressed

File: python/lib/sift_py/file_attachment/service.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            if not next_page_token:
+                break
+
+        return remote_files
+
+    def upload_attachment(
+        self,
+        path: Union[str, Path],
+        entity: Entity,
+        metadata: Optional[Metadata],
```
[the remark this answers is not in the document]
Comment (zachAtSiftStack) [reply python/lib/sift_py/file_attachment/service.py:67]: Addressed with = None

File: python/lib/sift_py/file_attachment/_internal/upload.py (hunk 2 above)
Comment (zachAtSiftStack) [reply python/lib/sift_py/file_attachment/_internal/upload.py]: Addressed with = None

**Pull request merged**